AnkiDroid/src/main/java/com/ichi2/libanki/Sched.java

```diff
@@ -1270,28 +1304,31 @@ private boolean _fillRev() {
         while (mRevDids.size() > 0) {
             long did = mRevDids.getFirst();
             int lim = Math.min(mQueueLimit, _deckRevLimit(did));
-            Cursor cur = null;
             if (lim != 0) {
                 mRevQueue.clear();
-                // fill the queue with the current did
-                try {
-                    cur = mCol
-                            .getDb()
-                            .getDatabase()
-                            .query(String.format("SELECT id " +
-                                    "FROM cards " +
-                                    "WHERE did = %s " +
-                                    "AND queue = 2 " +
-                                    "AND due <= %s " +
-                                    "ORDER BY ivl ASC " +
-                                    "LIMIT %s", did, mToday, lim), null);
-                    while (cur.moveToNext()) {
-                        mRevQueue.add(cur.getLong(0));
-                    }
-                } finally {
-                    if (cur != null && !cur.isClosed()) {
-                        cur.close();
-                    }
+                //don't care about any limit with young cards; just want them all mixed
+                List<Long> maybeYoungCards = getCardIdsWithQuery(String.format("SELECT id " +
+                        "FROM cards " +
+                        "WHERE did = %s " +
+                        "AND queue = 2 " +
+                        "AND due <= %s " +
+                        "AND ivl < 21", did, mToday));
+                if (maybeYoungCards.size() > 0) {
+                    mRevQueue.addAll(maybeYoungCards);
+                } else {
```

Comment:
So, the `if` / `else` here looks like we want to only populate the review queue with young cards if any exist. That seems reasonable to me. But if I blaze through all the young cards, will I get bounced back to the main deck screen before being able to review more mature cards? Not sure if my question is clear. Basically, I'm curious why there is this conditional here. It makes it look like mature cards will not be reviewed if you have any young cards. Again, this part seems reasonable. But it seems like not adding the mature cards means I would get bumped back to deck list and have to click back in to finish reviewing the mature cards. And that bump seems weird.

Do we exclude the mature cards here because of the later shuffling step? If that's the case, then this seems like the best solution for now (just get bumped back to main deck list after young cards and click back in to review the mature ones). Would be nice to just have it be seemless tho

Reply:
Good question -- you actually don't get kicked back, it just seamlessly switches to mature once you're done with young. Not sure exactly why the logic works that way but hey I'm not complaining

Reply:
And yeah exactly -- we want to shuffle but not shuffle them alongside mature as well (so that they're always prioritized before)